### ISSUE TITLE
compute_gene_trends throws ValueError: X data must not contain Inf nor NaN

### ISSUE THREAD

**Issue by widsquid** [dpeerlab/Palantir#98]
Thanks for creating this tool, it has been interesting so far. Sadly after much success I now get this error with the trend computation function. I have checked for NAs and Infs and zeros. I have tried to create a tester imp_df of random floats (no NAs nan Inf zeros) to input instead and still the same result. Any idea what could be the problem? I installed new from from git clone yesterday. All my data comes straight from 10x output. I am wondering could the NA be from the pr_res data? Thanks so much.

**gene_trends = palantir.presults.compute_gene_trends(pr_res, imp_df.loc[:, genes])**
---------------------------------------------------------------------------
_RemoteTraceback                          Traceback (most recent call last)
_RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 428, in _process_worker
    r = call_item()
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 275, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/_parallel_backends.py", line 620, in __call__
    return self.func(*args, **kwargs)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py", line 289, in __call__
    for func, args, kwargs in self.items]
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py", line 289, in <listcomp>
    for func, args, kwargs in self.items]
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/palantir/presults.py", line 162, in gam_fit_predict
    y_pred = gam.predict(pred_x)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/pygam.py", line 434, in predict
    return self.predict_mu(X)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/pygam.py", line 414, in predict_mu
    features=self.feature, verbose=self.verbose)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/utils.py", line 273, in check_X
    name='X data', verbose=verbose)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/utils.py", line 171, in check_array
    raise ValueError('{} must not contain Inf nor NaN'.format(name))
ValueError: X data must not contain Inf nor NaN
"""

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
/tmp/ipykernel_1341918/4064161686.py in <module>
----> 1 gene_trends = palantir.presults.compute_gene_trends(pr_res, imp_df.loc[:, genes])

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/palantir/presults.py in compute_gene_trends(pr_res, gene_exprs, lineages, n_splines, spline_order, n_jobs)
    120                 spline_order
    121             )
--> 122             for gene in gene_exprs.columns
    123         )
    124 

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py in __call__(self, iterable)
   1096 
   1097             with self._backend.retrieval_context():
-> 1098                 self.retrieve()
   1099             # Make sure that we get a last message telling us we are done
   1100             elapsed_time = time.time() - self._start_time

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py in retrieve(self)
    973             try:
    974                 if getattr(self._backend, 'supports_timeout', False):
--> 975                     self._output.extend(job.get(timeout=self.timeout))
    976                 else:
    977                     self._output.extend(job.get())

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/_parallel_backends.py in wrap_future_result(future, timeout)
    565         AsyncResults.get from multiprocessing."""
    566         try:
--> 567             return future.result(timeout=timeout)
    568         except CfTimeoutError as e:
    569             raise TimeoutError from e

~/anaconda3/envs/env_palantir/lib/python3.7/concurrent/futures/_base.py in result(self, timeout)
    433                 raise CancelledError()
    434             elif self._state == FINISHED:
--> 435                 return self.__get_result()
    436             else:
    437                 raise TimeoutError()

~/anaconda3/envs/env_palantir/lib/python3.7/concurrent/futures/_base.py in __get_result(self)
    382     def __get_result(self):
    383         if self._exception:
--> 384             raise self._exception
    385         else:
    386             return self._result

ValueError: X data must not contain Inf nor NaN

**Comment by ManuSetty** [comment]
Can you please run the function with `n_jobs=1`. This will help isolate where the issue is for better debugging.

**Comment by widsquid** [comment]
Thanks so much for the reply. Here is the output. Let me know if there is any other info that might help.

**gene_trends = palantir.presults.compute_gene_trends(pr_res, imp_df.loc[:, genes], n_jobs=1)**
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/ipykernel_1341918/2426816647.py in <module>
----> 1 gene_trends = palantir.presults.compute_gene_trends(pr_res, imp_df.loc[:, genes], n_jobs=1)

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/palantir/presults.py in compute_gene_trends(pr_res, gene_exprs, lineages, n_splines, spline_order, n_jobs)
    120                 spline_order
    121             )
--> 122             for gene in gene_exprs.columns
    123         )
    124 

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py in __call__(self, iterable)
   1083             # remaining jobs.
   1084             self._iterating = False
-> 1085             if self.dispatch_one_batch(iterator):
   1086                 self._iterating = self._original_iterator is not None
   1087 

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py in dispatch_one_batch(self, iterator)
    899                 return False
    900             else:
--> 901                 self._dispatch(tasks)
    902                 return True
    903 

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py in _dispatch(self, batch)
    817         with self._lock:
    818             job_idx = len(self._jobs)
--> 819             job = self._backend.apply_async(batch, callback=cb)
    820             # A job can complete so quickly than its callback is
    821             # called before we get here, causing self._jobs to

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/_parallel_backends.py in apply_async(self, func, callback)
    206     def apply_async(self, func, callback=None):
    207         """Schedule a func to be run"""
--> 208         result = ImmediateResult(func)
    209         if callback:
    210             callback(result)

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/_parallel_backends.py in __init__(self, batch)
    595         # Don't delay the application, to avoid keeping the input
    596         # arguments in memory
--> 597         self.results = batch()
    598 
    599     def get(self):

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py in __call__(self)
    287         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    288             return [func(*args, **kwargs)
--> 289                     for func, args, kwargs in self.items]
    290 
    291     def __reduce__(self):

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py in <listcomp>(.0)
    287         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    288             return [func(*args, **kwargs)
--> 289                     for func, args, kwargs in self.items]
    290 
    291     def __reduce__(self):

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/palantir/presults.py in gam_fit_predict(x, y, weights, pred_x, n_splines, spline_order)
    160     if pred_x is None:
    161         pred_x = x
--> 162     y_pred = gam.predict(pred_x)
    163 
    164     # Standard deviations

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/pygam.py in predict(self, X)
    432             containing predicted values under the model
    433         """
--> 434         return self.predict_mu(X)
    435 
    436     def _modelmat(self, X, term=-1):

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/pygam.py in predict_mu(self, X)
    412         X = check_X(X, n_feats=self.statistics_['m_features'],
    413                     edge_knots=self.edge_knots_, dtypes=self.dtype,
--> 414                     features=self.feature, verbose=self.verbose)
    415 
    416         lp = self._linear_predictor(X)

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/utils.py in check_X(X, n_feats, min_samples, edge_knots, dtypes, features, verbose)
    271     # basic diagnostics
    272     X = check_array(X, force_2d=True, n_feats=n_feats, min_samples=min_samples,
--> 273                     name='X data', verbose=verbose)
    274 
    275     # check our categorical data has no new categories

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/utils.py in check_array(array, force_2d, n_feats, ndim, min_samples, name, verbose)
    169     # check finite
    170     if not(np.isfinite(array).all()):
--> 171         raise ValueError('{} must not contain Inf nor NaN'.format(name))
    172 
    173     # check ndim

ValueError: X data must not contain Inf nor NaN

**Comment by ManuSetty** [comment]
Can you please check if there is NA or inf in `imp_df`. `imp_df.isna().sum()` and `imp_df.isinf().sum()`

**Comment by widsquid** [comment]
Sure thanks so much.

<img width="1398" alt="Screen Shot 2023-02-18 at 10 46 39 PM" src="https://user-images.githubusercontent.com/36773215/219933381-4c0abcb0-0675-4ddf-95ba-e925d0dcb69b.png">

<img width="1407" alt="Screen Shot 2023-02-18 at 10 46 23 PM" src="https://user-images.githubusercontent.com/36773215/219933373-1d0cbea4-53bc-4412-bdb9-42b14b3bda01.png">

**Comment by ManuSetty** [comment]
Can you please share the output of `pr_res` by running the `palantir.plot.plot_palantir_results`? Looks like it is an issue in there and not from imputation.

**Comment by widsquid** [comment]
<img width="1213" alt="Screen Shot 2023-02-21 at 1 36 55 AM" src="https://user-images.githubusercontent.com/36773215/220307328-33876d3d-4a34-4aeb-9823-7098ed8ecc48.png">

**Comment by ManuSetty** [comment]
Do any of `pr_res.pseudotime` or `pr_res.branch_probs` contain infinities or NA?

**Comment by widsquid** [comment]
It does not appear that way. Is this the correct way of determining that? Thanks again.
W
________________________________

**Comment by ManuSetty** [comment]
This is puzzling! Do you mind sharing a pickle object of `pr_res` and `imp_df`? I can examine it in my end to better understand this.

**Comment by widsquid** [comment]
[Archive.zip](https://github.com/dpeerlab/Palantir/files/10820752/Archive.zip)

**Comment by widsquid** [comment]
sure, thanks for your efforts. please let me know if you do not successfully receive the files.
W

**Comment by ManuSetty** [comment]
Sorry, can you please clarify what format the file is saved and how I can load it.

**Comment by widsquid** [comment]
Unzip to find two .pkl files I believe.
W

On Mar 4, 2023, at 8:49 PM, Manu Setty ***@***.***> wrote:

﻿

Sorry, can you please clarify what format the file is saved and how I can load it.

**Comment by widsquid** [comment]
were you able to successfully unzip the file? my mistake I should have said two .obj files saved using pkl

**Comment by ManuSetty** [comment]
Looks like the `pr_res` file might be corrupted. Can you please share again? 
<img width="666" alt="image" src="https://user-images.githubusercontent.com/2151404/224153963-e4d4fbbc-b1ab-4845-bb7f-e4854f0553f0.png">

**Comment by widsquid** [comment]
[https://res-geo.cdn.office.net/assets/mail/file-icon/png/generic_16x16.png]pr_res.obj<https://1drv.ms/u/s!Au68to3eRf-VgQllb57d_tBHrgMQ>
[https://res-geo.cdn.office.net/assets/mail/file-icon/png/generic_16x16.png]imp_df.obj<https://1drv.ms/u/s!Au68to3eRf-VgQpWY0tr_okzo3ym>


________________________________

**Comment by ManuSetty** [comment]
Still the same issue.
<img width="666" alt="image" src="https://user-images.githubusercontent.com/2151404/224233546-0f121573-fb60-4094-96a4-0286db27b6fe.png">


What version of pickle are you using to save?

**Comment by widsquid** [comment]
I tried to save the files again. Let me know if you can get them loaded. Thanks again for all the trouble, everything else looks great for the figures I am trying to make I except I would really like to have some of these plots in there as well. I hope we can get to the bottom of this.
[https://res-geo.cdn.office.net/assets/mail/file-icon/png/generic_16x16.png]imp_df.obj<https://1drv.ms/u/s!Au68to3eRf-VgQuTVwUbkpsnIBv0>
[https://res-geo.cdn.office.net/assets/mail/file-icon/png/generic_16x16.png]pr_res.obj<https://1drv.ms/u/s!Au68to3eRf-VgQxeDA9qYi_khK-8>

W
________________________________

**Comment by ManuSetty** [comment]
Sorry still the same issue
<img width="666" alt="image" src="https://user-images.githubusercontent.com/2151404/224520554-f2cb0331-760c-464f-bde1-3c67c3b957ab.png">


Perhaps you can save the information in there as text:
pr_res.pseudotime.to_csv('pt.csv')
pr_res.branch_probs.to_csv('branch_probs.csv')
pr_res.entropy.to_csv('entropy.csv')

**Comment by widsquid** [comment]
[https://res-geo.cdn.office.net/assets/mail/file-icon/png/csv_16x16.png]branch_probs.csv<https://1drv.ms/u/s!Au68to3eRf-VgQ3g0zf8xiFj29Fs>
[https://res-geo.cdn.office.net/assets/mail/file-icon/png/csv_16x16.png]entropy.csv<https://1drv.ms/u/s!Au68to3eRf-VgQ5AV0sHwxrf6r66>
[https://res-geo.cdn.office.net/assets/mail/file-icon/png/csv_16x16.png]pt.csv<https://1drv.ms/u/s!Au68to3eRf-VgQ_92woP8mm_u6Q4>
My apologies, certainly, here they are.
W
________________________________

**Comment by ManuSetty** [comment]
Can you please email to `manu.n.setty@gmail.com`? The above links are causing Onedrive to try and convert the files which never finishes.

**Comment by ManuSetty** [comment]
This looks an issue with a subset of the branches. If you use the `Peri` or `SRC` lineages, the trends will be computed without issues
<img width="865" alt="image" src="https://user-images.githubusercontent.com/2151404/224591154-7b76be98-83d4-4685-a8d0-49b7ad8f8399.png">


For the `Failed` lineage, the maximum probability of any given cell is 0.4. When assigning cells to branches, our code looks for cells with at least 0.7 probability (https://github.com/dpeerlab/Palantir/blob/4d4f9fcdbfebcf47b4ca1b7ac8a0fb1b8e5fc089/src/palantir/presults.py#L94). You will need to make a tweak in this line to get trends for this lineage.

**Comment by widsquid** [comment]
Amazing I can not wait try it. I will let you know if it works. Thanks so much.
W

On Mar 12, 2023, at 7:00 PM, Manu Setty ***@***.***> wrote:

﻿

This looks an issue with a subset of the branches. If you use the Peri or SRC lineages, the trends will be computed without issues
[image]<https://user-images.githubusercontent.com/2151404/224591154-7b76be98-83d4-4685-a8d0-49b7ad8f8399.png>

For the Failed lineage, the maximum probability of any given cell is 0.4. When assigning cells to branches, our code looks for cells with at least 0.7 probability (https://github.com/dpeerlab/Palantir/blob/4d4f9fcdbfebcf47b4ca1b7ac8a0fb1b8e5fc089/src/palantir/presults.py#L94). You will need to make a tweak in this line to get trends for this lineage.

**Comment by widsquid** [comment]
Same result if I modify the threshold to 0.0. Would that be typical in a case like this? Also what does it say about this particular lineage if it is not calling high probability cells?

**Comment by ManuSetty** [comment]
It is puzzling that lowering the threshold did not work - are you seeing the error at the same code location?  

Low probability on a lineage indicates that it might not be a true terminal state.

**Comment by widsquid** [comment]
there is a possibility it is not a true terminal state, perhaps rather a failed trajectory. There are however interesting differentially expressed genes in that region.  Is it still possible to plot those on a trajectory? I suppose perhaps lowering the threshold should allow this if working?

This is the error, seems like the same outcome at least (of NA or undefined values being generated):



_RemoteTraceback                          Traceback (most recent call last)
_RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 428, in _process_worker
    r = call_item()
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/externals/loky/process_executor.py", line 275, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/_parallel_backends.py", line 620, in __call__
    return self.func(*args, **kwargs)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py", line 289, in __call__
    for func, args, kwargs in self.items]
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py", line 289, in <listcomp>
    for func, args, kwargs in self.items]
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/palantir/presults.py", line 162, in gam_fit_predict
    y_pred = gam.predict(pred_x)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/pygam.py", line 434, in predict
    return self.predict_mu(X)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/pygam.py", line 414, in predict_mu
    features=self.feature, verbose=self.verbose)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/utils.py", line 273, in check_X
    name='X data', verbose=verbose)
  File "/home/w/anaconda3/envs/env_palantir/lib/python3.7/site-packages/pygam/utils.py", line 171, in check_array
    raise ValueError('{} must not contain Inf nor NaN'.format(name))
ValueError: X data must not contain Inf nor NaN
"""

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
/tmp/ipykernel_2663296/3306120243.py in <module>
      3 imp_df = pd.DataFrame(ad[:, genes].layers['MAGIC_imputed_data'],
      4                      index=ad.obs_names, columns=genes)
----> 5 gene_trends = palantir.presults.compute_gene_trends( pr_res, imp_df.loc[:, genes])#, lineages = ['SRC'], n_jobs = 1)
      6 #gene_trends = palantir.presults.compute_gene_trends( pr_res, imp_df.loc[:, genes], n_jobs = 1)

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/palantir/presults.py in compute_gene_trends(pr_res, gene_exprs, lineages, n_splines, spline_order, n_jobs)
    120                 spline_order
    121             )
--> 122             for gene in gene_exprs.columns
    123         )
    124 

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py in __call__(self, iterable)
   1096 
   1097             with self._backend.retrieval_context():
-> 1098                 self.retrieve()
   1099             # Make sure that we get a last message telling us we are done
   1100             elapsed_time = time.time() - self._start_time

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/parallel.py in retrieve(self)
    973             try:
    974                 if getattr(self._backend, 'supports_timeout', False):
--> 975                     self._output.extend(job.get(timeout=self.timeout))
    976                 else:
    977                     self._output.extend(job.get())

~/anaconda3/envs/env_palantir/lib/python3.7/site-packages/joblib/_parallel_backends.py in wrap_future_result(future, timeout)
    565         AsyncResults.get from multiprocessing."""
    566         try:
--> 567             return future.result(timeout=timeout)
    568         except CfTimeoutError as e:
    569             raise TimeoutError from e

~/anaconda3/envs/env_palantir/lib/python3.7/concurrent/futures/_base.py in result(self, timeout)
    433                 raise CancelledError()
    434             elif self._state == FINISHED:
--> 435                 return self.__get_result()
    436             else:
    437                 raise TimeoutError()

~/anaconda3/envs/env_palantir/lib/python3.7/concurrent/futures/_base.py in __get_result(self)
    382     def __get_result(self):
    383         if self._exception:
--> 384             raise self._exception
    385         else:
    386             return self._result

ValueError: X data must not contain Inf nor NaN

**Comment by widsquid** [comment]
Hello again, sorry for the delay. I was able to get back to it and I have run the script again I believe with the modified threshold. Can we assume that the threshold is set to zero based on running the command in my attached image and seeing the 0.0 in that line? Thanks again.
<img width="960" alt="Screen Shot 2023-04-22 at 7 55 06 PM" src="https://user-images.githubusercontent.com/36773215/233812396-7b89274c-2a8e-4fdd-b4af-415e2b764063.png">
<img width="1038" alt="Screen Shot 2023-04-22 at 7 54 48 PM" src="https://user-images.githubusercontent.com/36773215/233812413-c2f8d39f-931d-4ab2-9f92-aa4b73c654fc.png">

**Comment by widsquid** [comment]
<img width="727" alt="Screen Shot 2023-04-22 at 8 41 11 PM" src="https://user-images.githubusercontent.com/36773215/233813395-a9de8bfb-6ce1-4810-819d-b92d290f53d3.png">
<img width="723" alt="Screen Shot 2023-04-22 at 8 43 41 PM" src="https://user-images.githubusercontent.com/36773215/233813401-3b2d2481-d810-4823-a913-cad2b71967cd.png">

**Comment by ManuSetty** [comment]
This should have worked. I can investigate this on my end - can you please reshape the files again? Sorry, I cannot seem to locate them.

**Comment by widsquid** [comment]
Certainly, my pleasure, thanks for taking a look.
W

**Comment by widsquid** [comment]
[https://res-h3.public.cdn.office.net/assets/mail/file-icon/png/generic_16x16.png]pr_res.obj<https://1drv.ms/u/s!Au68to3eRf-VgSCjrJrn3-xmPvSc>
[https://res.cdn.office.net/assets/mail/file-icon/png/csv_16x16.png]entropy.csv<https://1drv.ms/u/s!Au68to3eRf-VgSH4vvIIahbL9OpF>
[https://res.cdn.office.net/assets/mail/file-icon/png/csv_16x16.png]branch_probs.csv<https://1drv.ms/u/s!Au68to3eRf-VgSICvCrLNuKza4Sb>
[https://res-h3.public.cdn.office.net/assets/mail/file-icon/png/generic_16x16.png]imp_df.obj<https://1drv.ms/u/s!Au68to3eRf-VgSPrSsrskMTAMixn>
[https://res.cdn.office.net/assets/mail/file-icon/png/csv_16x16.png]pt.csv<https://1drv.ms/u/s!Au68to3eRf-VgSTuS6aOaJLAsc8u>

Are these all the files you need? Thanks.
W
________________________________

**Comment by widsquid** [comment]
FYI when computing gene trends, "SRC13" trajectory works alone, "Peri" works alone but they do not work together and "Failed" will not work alone or with any combination.

**Comment by ManuSetty** [comment]
The links dont seem to be working - can you please me again `manu.n.setty <AT> gmail.com `

**Comment by widsquid** [comment]
sorry, yes no problem!
W

On Apr 26, 2023, at 4:00 PM, Manu Setty ***@***.***> wrote:

﻿

The links dont seem to be working - can you please me again manu.n.setty <AT> gmail.com

**Comment by ManuSetty** [comment]
Hello again 

The changes we discussed did work for me and the gene trend computation was successful for all lineage and genes you shared with me 

<img width="882" alt="image" src="https://user-images.githubusercontent.com/2151404/235408863-431e794a-1853-42de-a409-03bd4784ecdc.png">

**Comment by widsquid** [comment]
Ok that is interesting. Any thoughts as to why it might be returning that error on my end? I could try a fresh install of the package?
W
________________________________

**Comment by ManuSetty** [comment]
Yes - please pull the latest repo, make the changes and then reinstall. Please try starting your notebook server only after re-installing the package.

**Comment by widsquid** [comment]
Some progress? After a reinstall and open a new notebook I got a second trajectory on the trend plots but it ignored the third one.

['SRC13', 'Peri', 'Failed']
<img width="730" alt="Screen Shot 2023-05-01 at 5 47 33 PM" src="https://user-images.githubusercontent.com/36773215/235537423-4f89b5a8-e0ab-48d3-8bfc-89af168958ba.png">
<img width="651" alt="Screen Shot 2023-05-01 at 5 01 26 PM" src="https://user-images.githubusercontent.com/36773215/235537449-ec8bd4ae-cf58-4866-b775-4ab3ff47d8a6.png">

**Comment by ManuSetty** [comment]
Does the gene trends object not contain the results for the other lineages?

Manu

**Comment by widsquid** [comment]
No good point, it does not. Looks like just for the two that are plotted, trends and std each (49 genes x 500 waypoints).

**Comment by widsquid** [comment]
I sent some information to your gmail. Thanks again.
W

**Comment by widsquid** [comment]
Thank you, with the clues you provided I was able to work out where the problem was coming from. I tested a number of terminal cell combinations and and thresholds and determined that with certain combinations of terminal cells, the probabilities were indeed dropping well below the threshold. We were also able to reproduce this error in another dataset that had previously resolved four trajectories by simply setting an additional terminal cell too close to one of the existing terminal cells, thereby reducing the confidence of the existing trajectory to below the threshold I suppose. In any case thank you for your efforts.
